grpc-client/src/main/scala/com/avast/cactus/grpc/client/ClientMacros.scala

```diff
@@ -0,0 +1,172 @@
+package com.avast.cactus.grpc.client
+
+import com.avast.cactus.CactusMacros
+import com.avast.cactus.grpc.ServerError
+import com.google.common.util.concurrent.ListenableFuture
+import com.google.protobuf.MessageLite
+import io.grpc.stub.AbstractStub
+
+import scala.reflect.macros.whitebox
+
+class ClientMacros(val c: whitebox.Context) {
+
+  import c.universe._
+
+  def mapClientToTraitWithInterceptors[GrpcClientStub <: AbstractStub[GrpcClientStub]: WeakTypeTag, MyTrait: WeakTypeTag](
+      interceptors: c.Tree*)(ec: c.Tree, ex: c.Tree): c.Expr[MyTrait] = {
+
+    val stubType = weakTypeOf[GrpcClientStub]
+    val traitType = weakTypeOf[MyTrait]
+    val traitTypeSymbol = traitType.typeSymbol
+
+    if (!traitTypeSymbol.isClass || !traitTypeSymbol.asClass.isTrait || traitTypeSymbol.typeSignature.takesTypeArgs) {
+      c.abort(c.enclosingPosition, s"The $traitTypeSymbol must be a plain trait without type arguments")
+    }
+
+    val apiMethods = getApiMethods(stubType)
+    val methodsMappings = getMethodsMapping(traitType, stubType, apiMethods)
+
+    checkAbstractMethods(traitType, methodsMappings.keySet)
+
+    val channelVar = CactusMacros.getVariable(c)
+    val mappingMethods = methodsMappings.map { case (im, am) => generateMappingMethod(im, am) }
+
+    val stub = {
+      q" ${stubType.typeSymbol.owner}.newFutureStub($channelVar).withInterceptors(com.avast.cactus.grpc.client.ClientMetadataInterceptor) "
+    }
+
+    c.Expr[MyTrait] {
+      q"""
+         new com.avast.cactus.grpc.client.ClientInterceptorsWrapper(scala.collection.immutable.Seq(..$interceptors)) with $traitType {
+            private val ex: java.util.concurrent.Executor = $ex
+
+            private val stub = $stub
+
+            import com.avast.cactus.grpc.client.ClientCommonMethods._
+            import com.avast.cactus.v3._
+
+            ..$mappingMethods
+        }
+       """
+    }
+  }
+
+  private def generateMappingMethod(implMethod: ImplMethod, apiMethod: ApiMethod): Tree = {
+    q"""
+       override def ${implMethod.name}(request: ${implMethod.request}): scala.concurrent.Future[com.avast.cactus.grpc.ServerResponse[${implMethod.response}]] = withInterceptors {
+          request.asGpb[${apiMethod.request}] match {
+             case org.scalactic.Good(req) => executeRequest[${apiMethod.request}, ${apiMethod.response}, ${implMethod.response}](req, stub.${apiMethod.name}, ex)
```

Comment:
Just out of curiosity, is there a reason to use both `scalactic` and `cats` in one library?

Reply:
I am thinking about changing the API to `Either` from `scalactic.Or`, as well as compiling only for Scala 2.12. That change could mean also removing the Scalactic completely if I will implement the errors accumulation with Cats.
However that's a subject to another PR.